pyanaconda/ui/tui/hubs/__init__.py

```diff
@@ -35,49 +40,56 @@ class TUIHub(TUIObject, common.Hub):
        :param categories: list all the spoke categories to be displayed in this Hub
        :type categories: list of strings
 
-       :param title: title for this Hub
-       :type title: str
+
 
        .. inheritance-diagram:: TUIHub
           :parts: 3
     """
 
     categories = []
-    title = N_("Default HUB title")
 
-    def __init__(self, app, data, storage, payload, instclass):
-        TUIObject.__init__(self, app, data)
+    def __init__(self, data, storage, payload, instclass):
+        TUIObject.__init__(self, data)
         common.Hub.__init__(self, storage, payload, instclass)
+        self.title = N_("Default HUB title")
+        self._container = None
 
         self._spokes = {}     # holds spokes referenced by their class name
-        self._keys = {}       # holds spokes referenced by their user input key
         self._spoke_count = 0
 
-    def setup(self, environment="anaconda"):
+        # we want user input
+        self.input_required = True
+
+    def setup(self, args="anaconda"):
+        environment = args
         cats_and_spokes = self._collectCategoriesAndSpokes()
         categories = cats_and_spokes.keys()
 
-        for c in sorted(categories, key=lambda c: c.title):
+        for c in sorted(categories, key=lambda i: i.title):
+
+            hub_spokes = []
+            for spoke_class in cats_and_spokes[c]:
+                # Do the checks for the spoke and create the spoke
+                if spoke_class.should_run(environment, self.data):
+                    spoke = spoke_class(self.data, self.storage, self.payload, self.instclass)
 
-            for spokeClass in sorted(cats_and_spokes[c], key=lambda s: s.title):
-                # Check if this spoke is to be shown in anaconda
-                if not spokeClass.should_run(environment, self.data):
-                    continue
+                    if spoke.showable:
+                        spoke.initialize()
+                    else:
+                        log.warning("Spoke %s initialization failure!", spoke.__class__.__name__)
+                        del spoke
+                        continue
 
-                spoke = spokeClass(self.app, self.data, self.storage, self.payload, self.instclass)
+                    if spoke.indirect:
+                        continue
 
-                if spoke.showable:
-                    spoke.initialize()
-                else:
-                    del spoke
-                    continue
+                    hub_spokes.append(spoke)
 
-                if spoke.indirect:
-                    continue
+            # sort created spokes and add them to result structures
+            for spoke in sorted(hub_spokes, key=lambda s: s.title):
 
                 self._spoke_count += 1
-                self._keys[self._spoke_count] = spoke
-                self._spokes[spokeClass.__name__] = spoke
+                self._spokes[spoke.__class__.__name__] = spoke
```

Comment:
I doesn't, however, it is working more or less on a good word right now. The problem is, that I missed the ordering of the keys dictionary before which wasn't quite bullet proof. I'll try to use some new solution for this.